### ISSUE TITLE
Cache/buffer for listening to http/https radio streams?

### ISSUE THREAD

**Issue by bikubi** [MusicPlayerDaemon/MPD#1063]
## Question

I use mpd to listen to web radio streams (mp3 and aac over http and https). Every couple of hours, I hear stutters. I have to stop & start the stream again -- I presume, to refill the cache.

From what I can tell:

- mpd's cache/buffer is very short since streams start almost instantaneously
- there is no way to set the buffer/cache size, like mplayer's `-cache` and `-cache-min` parameters or VLC's `network-caching`
- `input_cache` does not affect this & never will?
- the `curl` plugins handles "fetching" the stream, other plugins (ffmpeg/mpg123) decode it?

Are my assumptions correct?  
Is there anything in development, or planned, that would improve this?  
Or am I using the wrong tool?

Ideas:

- tell the curl plugin to cache/buffer, somehow (where would I start to address this: curl plugin? ffmpeg?)
- use a http proxy to pre-cache, somehow

To me, these seem far fetched. Any input & other ideas, pointers, corrections greatly appreciated!

**Comment by MaxKellermann** [comment]
What you observe happens because the clock of your radio stream and the clock of your DAC aren't 100% perfect, and they diverge a tiny bit. In your case, your DAC is faster than the radio, so MPD's buffered data will get smaller and smaller, until it runs empty after hours.
Increasing the buffer will only postpone this problem to a few more hours, but will not fix this completely. Caching/buffering cannot solve the problem.
There may be two approaches to fix this:
- let MPD adjust the playback speed automatically to keep the buffer at a constant level (which is difficult to get right, because MPD doesn't know why data doesn't come quickly enough - networking problems or clock skew? - and it will reduce sound quality)
- if the buffer runs empty, don't stutter many times a second, but instead let MPD wait for a second or so until the buffer is filled again

Of course, the opposite can happen: the radio stream is faster than your DAC. This needs similar measures to prevent problems.
It is true that MPD is pretty bad at this currently, and needs to improve.

**Comment by bikubi** [comment]
Thank you @MaxKellermann for the fascinating insight. Wasn't aware of clock skew in this field. I'd like to add

* when I stop & restart after stuttering, often the stuttering comes back *right after*. If I understand correctly, this shouldn't be due to clock skew (which accumulates over time), but due to network problems.
* the stutters, btw, don't happen "many times a second", fortunately, but in random intervals between maybe 2 and 10 seconds. Still pretty annoying :)
* Even if buffering/caching wouldn't *solve* the issue, wouldn't it improve the situation? I rarely listen to one stream for more than 18h. If I do (leave it on over night / indefinitely), I wouldn't mind *replenishing* the cache, say, daily.
* No other player I have used starts playback as quick as mpd. When I used to use mplayer, I would configure the cache to hold up to 10 seconds; with mpv I do not anymore (it might have better anti-clockskew/underrun slow/down/up strategies). Browsers (`<audio>`) are the worst offenders with ~5s of cache/buffer that you have no control over.
* In other words: I, personally, would prefer an imperfect mitigation (which helps with my bad wifi "for a few more hours") over a perfect anti-clock-skew solution. I would also prefer a break/fadeout/whatever over stutter.
* I might be missing things & just trying to add observations & feedback. Please don't read this as nitpicking or entitlement, I am grateful for mpd :)
